cmake/toolchain/llvm/target.cmake

```diff
@@ -1,3 +1,25 @@
 # SPDX-License-Identifier: Apache-2.0
 
-# Intentionally left blank.
+if(CONFIG_LLVM_USE_LD)
+  set(LINKER ld)
+elseif(CONFIG_LLVM_USE_LLD)
+  set(LINKER lld)
+else()
+  message(FATAL_ERROR
+          "Need to enable either CONFIG_LLVM_USE_LD or CONFIG_LLVM_USE_LLD!")
```

Comment:
> Is this really needed ? A choice will always have a value,...

Unless when editing the .config file manually or generating from fragments and some non-Zephyr script?

> so the only way that both CONFIG_LLVM_USE_LD and CONFIG_LLVM_USE_LLD could n is in case someone extended the choice with an extra config option.

Could someone really add an extra KConfig option without editing this code?

> Finally, fatal errors should be used with absolute caution when it comes to Kconfig settings because of #9573

I only skimmed 9573 sorry but the main message seems to be "don't generate invalid .config files that menuconfig" can't read again. It will of course always be possible to make invalid .config edits manually.

Can this FATAL_ERROR really cause invalid .config files to be generated?

In my experience CMake and Kconfig errors tend to be a nightmare to debug + no one wants to go anywhere near them. Very understandable when you look at major design mistakes like COMMAND_ERROR_IS_FATAL false by default, MYPROGRAM_NOT-FOUND and other crazy and super time-consuming CMake stuff like https://samthursfield.wordpress.com/2015/11/21/cmake-dependencies-between-targets-and-files-and-custom-commands/

So in general we need MORE user-friendly error messages, not fewer but maybe this particular case is different? Not convinced of that yet.

Reply:
If someone adds a new linker kconfig, this acts as a reminder that this also needs to be changed. I will change it to a warning instead.

Reply:
> Unless when editing the .config file manually or generating from fragments and some non-Zephyr script?

No, it will always be set, try this for yourself:
```
$ cat build/zephyr/.config |grep LLVM_USE
CONFIG_LLVM_USE_LD=y
# CONFIG_LLVM_USE_LLD is not set
$ sed 's/\(CONFIG_LLVM_USE_LD\)=y/# \1 is not set/g' build/zephyr/.config > tmp; mv tmp build/zephyr/.config
$ cat build/zephyr/.config |grep LLVM_USE
# CONFIG_LLVM_USE_LD is not set
# CONFIG_LLVM_USE_LLD is not set

.... Let's do the build:
$ ninja -Cbuild
ninja: Entering directory `build'
[0/1] Re-running CMake...
Loading Zephyr default modules (Zephyr base (cached)).
....
$ cat build/zephyr/.config |grep LLVM_USE
CONFIG_LLVM_USE_LD=y
# CONFIG_LLVM_USE_LLD is not set
```

As you see, touching `.config` will cause a CMake re-run which triggers a kconfig run that ensures all choices has a value.

> Could someone really add an extra KConfig option without editing this code?

You can extend choices with new entries downstream, for example like it's done here:
https://github.com/nrfconnect/sdk-nrf/blob/5f58fc4ae964e24e5b432bcfcca865f857286f62/Kconfig.nrf#L23-L38
where the Zephyr provided options for an mbedTLS implementation is extended with an extra choice entry in a downstream project.

> Can this FATAL_ERROR really cause invalid .config files to be generated?

No, but a `FATAL_ERROR` in CMake will prevent running `menuconfig`, cause the build file (which contains the `menuconfig` build target) has never been created.

Therefore, mis-configured Kconfig files should not result in fatal errors in CMake cause users will not be able to open `menuconfig` to fix the error.

So in this case, there is no value in throwing an error that a user cannot fix.
Plus it prevents downstream projects to provide extra choice options, and that's making it harder to extend Zephyr on a general basis.

In this particular case, there is probably little risk, but users will copy this pattern else-where, not realizing the damage / limitations it may introduce at other places.